### PR TITLE
[WIP] Add link to apiman console

### PR DESCRIPTION
**Pull request by gashcrumb**
This PR adds a link to the [APIman console](http://www.apiman.io/latest/) when it's deployed in a project.

It's a WIP though for a few reasons:

1)  **Link location** - Currently the code adds it to the main navigation when the apiman service is present *however* maybe there's a better place for it.  For example, it'd be nice to update the link that's present in the overview that shows up due to the route.
2)  **Link title** - `API Management` or `Manage APIs` is too wordy, `APIman` isn't appropriate either really.  I'm going to ask the apiman folks to chime on this a bit too actually, but any thoughts appreciated.
3)  **icon** - went with a generic wrench for now.  We annotate the template and service I think in fabric8 with an icon path, so could use that, or I could use a data url and encode the apiman icon, think that might look out of place though.

When you click the link and it takes you over to the apiman console there will be a link back to the openshift console, think slightly similar to the java console.  Here's a screenshot of it's current state:

apiman

@jwforres FYI

### CODE REVIEW

**Comment by gashcrumb** [comment]
Just got the updated apiman header installed, here's where you go when you click the link:

apiman1

**Comment by danmcp** [comment]
Implements: https://trello.com/c/dVMWSrhL

**Comment by jwforres** [comment]
Are we expecting there to typically be one instance of api manager running in an openshift cluster?  Some of my feedback would change significantly based on the answer.

cc @smarterclayton

**Comment by gashcrumb** [comment]
I believe it's per-namespace/project, do I have that right @KurtStam?

**Comment by KurtStam** [comment]
Yes per namespace.

On Thu, Oct 1, 2015 at 1:22 PM, Stan Lewis <notifications@github.com> wrote:

> I believe it's per-namespace/project, do I have that right @KurtStam
> <https://github.com/KurtStam>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4878#issuecomment-144792623>.
>



-- 
Kurt T. Stam

twitter: @KurtStam
google+: kurt.stam@gmail.com

**Comment by KurtStam** [comment]
One per namespace.

**Comment by jwforres** [comment]
While I actually like the idea of it appearing in the left nav, I'm very concerned that its fetching the routes every time the nav gets rendered.  That means every time you change pages at least.  A watch (with a call to unwatch on destroy) would be a little better, that way it won't ever cause an extra call on pages where routes are already being watched.  But if we are going to do it this way then we probably should not include this in base openshift and ship this as a separate extension (and use the new extension mechanism to load arbitrary JS).

The downside to not doing it in the left nav and only changing the link on the apiman route would be that you can only get to it from the overview. That said, we should probably change the apiman route either way so that they don't jump over to that console without the token and back link.  I feel like this is going to be a common pattern that there is a default page you want to go to when you link to a route. @smarterclayton @pweil what if we define an annotation for this? openshift.io/route.default.link  and in the value we can let people specify {token} and {backLink} and the UI will replace those automatically.

As far as an icon, a wrench seems like a reasonable option to me.  @serenamarie125 see question number 3 in the PR description, let us know if there is a different icon you know of that we should be using.

For the link name, I actually like Manage APIs but it does look like we have a wrapping problem, @sg00dwin i'm guessing we have a fixed width set on the left nav, or is it min-width and we just need to add a nowrap?

**Comment by gashcrumb** [comment]
@jwforres can see what other options there are for showing/hiding that nav item.  Wonder, do you have an example of the type of watch you're talking about elsewhere in the console?

Definite +1 for an annotation that defines the initial path and some method of injecting the token/link back, we're only going to see more of a need for this as time goes on :-)

**Comment by jwforres** [comment]
overview.js does tons of them.  watch opens at most one websocket per
resource type and any time we get updates from the server it fires all
callbacks that are watching that resource type.

example:
https://github.com/openshift/origin/blob/master/assets/app/scripts/controllers/overview.js#L278
the details for what params it takes:
https://github.com/openshift/origin/blob/master/assets/app/scripts/services/data.js#L363

**Comment by smarterclayton** [comment]
Annotations on routes are an excellent idea


On Oct 1, 2015, at 9:32 PM, Jessica Forrester <notifications@github.com>
wrote:

While I actually like the idea of it appearing in the left nav, I'm very
concerned that its fetching the routes every time the nav gets rendered.
That means every time you change pages at least. A watch (with a call to
unwatch on destroy) would be a little better, that way it won't ever cause
an extra call on pages where routes are already being watched. But if we
are going to do it this way then we probably should not include this in
base openshift and ship this as a separate extension (and use the new
extension mechanism to load arbitrary JS).

The downside to not doing it in the left nav and only changing the link on
the apiman route would be that you can only get to it from the overview.
That said, we should probably change the apiman route either way so that
they don't jump over to that console without the token and back link. I
feel like this is going to be a common pattern that there is a default page
you want to go to when you link to a route. @smarterclayton
<https://github.com/smarterclayton> @pweil <https://github.com/pweil> what
if we define an annotation for this? openshift.io/route.default.link and in
the value we can let people specify {token} and {backLink} and the UI will
replace those automatically.

As far as an icon, a wrench seems like a reasonable option to me.
@serenamarie125 <https://github.com/serenamarie125> see question number 3
in the PR description, let us know if there is a different icon you know of
that we should be using.

For the link name, I actually like Manage APIs but it does look like we
have a wrapping problem, @sg00dwin <https://github.com/sg00dwin> i'm
guessing we have a fixed width set on the left nav, or is it min-width and
we just need to add a nowrap?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4878#issuecomment-144824466>.

**Comment by liggitt** [comment]
We shouldn't send the user's API token to an arbitrary URL based on data in an annotation.

**Comment by smarterclayton** [comment]
Which means third parties need to do an oauth dance with us.  What's the
scope of work to enable that?

On Oct 1, 2015, at 10:55 PM, Jordan Liggitt <notifications@github.com>
wrote:

We shouldn't send the user's API token to an arbitrary URL based on data in
an annotation.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4878#issuecomment-144845551>.

**Comment by liggitt** [comment]
> What's the scope of work to enable that?

Add grant approval and token scoping, probably. not small. I'm ok with going to the API proxy with a token... are we accessing a service to use this?

**Comment by gashcrumb** [comment]
@KurtStam, what's the token required for again for the apiman frontend?

**Comment by smarterclayton** [comment]
If we had macaroons we could solve it....

On Oct 1, 2015, at 11:01 PM, Jordan Liggitt <notifications@github.com>
wrote:

What's the scope of work to enable that?

Add grant approval and token scoping, probably. not small. I'm ok with
going to the API proxy with a token... are we accessing a service to use
this?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4878#issuecomment-144847109>.

**Comment by KurtStam** [comment]
@gashcrumb - the Apiman UI runs in the browser of the Hawtio/Openshift user, and uses the Apiman REST endpoint. So the REST endpoint needs to know what user is making the calls and that this user has a valid session. This is endpoint looks for the token on the request and then checks with Openshift for the validity of the token.

**Comment by KurtStam** [comment]
I wish Openshift used Keycloak so we don't have to reinvent the wheel on this. APIMan uses keycloak in the community project for all of this SSO stuff.

**Comment by gashcrumb** [comment]
@jwforres Reworked the code a bit so routes aren't fetched whenever the navbar redraws.  

Not sure what to do about the access token though, apiman definitely needs to see who the current user is and also needs access to the API server for some operations, so maybe this bit requires further discussion...

**Comment by smarterclayton** [comment]
We could whitelist routes we'll pass tokens to (vs just click through).  I
don't see how keycloak would allow this any more than our impl - the oauth
client still has to be registered ahead of time.

On Oct 1, 2015, at 5:01 PM, Jordan Liggitt <notifications@github.com> wrote:

What's the scope of work to enable that?

Add grant approval and token scoping, probably. not small. I'm ok with
going to the API proxy with a token... are we accessing a service to use
this?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4878#issuecomment-144847109>.

**Comment by gashcrumb** [comment]
@jwforres any word on this one?

**Comment by jwforres** [comment]
@spadgett

**Comment by jwforres** [comment]
cc @jhadvig

**Comment by gashcrumb** [comment]
Oh sorry, let's close this until some other things are ready like [this card](https://trello.com/c/Wrnv3JQj/563-8-whitelist-consoles-that-can-have-the-user-s-token-passed-to-them-from-the-openshift-console).
